cli/pcluster/utils.py

```diff
@@ -199,3 +200,130 @@ def get_supported_schedulers():
     :return: a tuple of strings of the supported scheduler
     """
     return "sge", "torque", "slurm", "awsbatch"
+
+
+def next_power_of_2(x):
+    """Given a number returns the following power of 2 of that number."""
+    return 1 if x == 0 else 2 ** (x - 1).bit_length()
+
+
+def get_subnet_cidr(vpc_cidr, occupied_cidr, min_subnet_size):
+    """
+    Decide the parallelcluster subnet size of the compute fleet.
+
+    :param vpc_cidr: the vpc_cidr in which the suitable subnet should be
+    :param occupied_cidr: a list of cidr of the already occupied subnets in the vpc
+    :param min_subnet_size: the minimum size of the subnet
+    :return:
+    """
+    default_target_size = 4000
+    target_size = max(default_target_size, 2 * min_subnet_size)
+    cidr = evaluate_cidr(vpc_cidr, occupied_cidr, target_size)
+    while cidr is None:
+        if target_size < min_subnet_size:
+            return None
+        target_size = target_size // 2
+        cidr = evaluate_cidr(vpc_cidr, occupied_cidr, target_size)
+    return cidr
+
+
+def evaluate_cidr(vpc_cidr, occupied_cidr, target_size):
+    """
+    Decide the first smallest suitable CIDR for a subnet with size >= target_size.
+
+    :param vpc_cidr: the vpc_cidr in which the suitable subnet should be
+    :param occupied_cidr: a list of cidr of the already occupied subnets in the vpc
+    :param target_size: the minimum target size of the subnet
+    :return: the suitable CIDR if found, else None
+    """
+    subnet_size, target_bitmask_length = _evaluate_subnet_size(target_size)
+    vpc_begin_address_decimal, vpc_end_address_decimal = _get_cidr_limits_as_decimal(vpc_cidr)
+
+    if vpc_end_address_decimal - vpc_begin_address_decimal + 1 < subnet_size:  # if we do not have enough space
+        return None
+
+    if not occupied_cidr:  # if we have space and no occupied cidr
+        return _decimal_ip_limits_to_cidr(vpc_begin_address_decimal, vpc_begin_address_decimal + subnet_size)
+
+    occupied_cidr_max_bitmask = max([get_bitmask(subnet_cidr) for subnet_cidr in occupied_cidr])
+    if occupied_cidr_max_bitmask > target_bitmask_length:
+        # This means that it's smaller, so we need to make it bigger
+        occupied_cidr = _expand_cidrs(occupied_cidr, min_size=target_bitmask_length)
+
+    # subnets_number is a list of pair(begin ip, end ip) obtained from the cidr. So for example
+    # 10.0.0.0/17 = 10.0.0.0, 10.0.127.255
+    begin_ip_index = 0
+    end_ip_index = 1
+    subnets_limits = [_get_cidr_limits_as_decimal(subnet) for subnet in occupied_cidr]
+    subnets_limits.sort(key=lambda x: x[1])  # sort by ending numbers, sorting by beginning is the same
+    # to check for space between the last occupied and the end of the vpc
+    subnets_limits.append((vpc_end_address_decimal, vpc_end_address_decimal))
+
+    if (subnets_limits[0][begin_ip_index] - vpc_begin_address_decimal) >= subnet_size:
```

Comment:
I think it is a best practice to have some blockers (returns) at function start but we could create an if/else block for the algorithmic part, since it doesn't have to many nested ifs